### ISSUE TITLE
Some questions about wannierized velocity matrix elements

### ISSUE THREAD

**Issue by ciccarino** [shankar1729/jdftx#180]
Hi Shankar,

I hope you are doing well. 

I had a few questions about the wannierized velocity matrix elements, specifically the off-diagonal terms. There are two points where I'm confused, the first being about apparent discontinuities in the elements as a function of wavevector, and the second about obeying symmetry. 

For all cases, I am reading in and transforming the matrix elements in a way similar to the [tutorial](http://jdftx.org/EphMatrixElements.html), except my einsum looks like this: `v = -1.0j*np.einsum("kba,kpbc,kcd->kpad", U.conj(), P, U)`. I added the `-1.0j` part based on line 295 of `WannierMinimizer_save.cpp`, but let me know if this is an incorrect interpretation. For any system with non-collinear spins I also set the datatype on reading in to `np.complex128`. 

For the first question about discontinuities, I will show here the off-diagonal matrix elements for cartesian direction x and n = 0, n' = 2 for the case of Aluminum. 
![velocities](https://user-images.githubusercontent.com/32557266/131522350-50d02455-8975-40ca-87e2-ba67c1f7d842.png)
The plots from left to right then show the real, imaginary and absolute value of the velocity matrix elements for a sampling of kpoints along the [0,0,0] -> [+/-0.5, 0, 0] path. The different rows correspond to different kpoint samplings. There are two lines in each plot, one corresponding to the -> [+0.5, 0, 0] kpoint path, and the other corresponding to the -> [-0.5, 0, 0] kpoint path. Since Aluminum is inversion symmetric, these paths should be equivalent, at least from an energetic standpoint (more on this below). 

My first question is related to the spikey nature of these data, in particular for the real and imaginary parts. How should I interpret this? I thought perhaps it was a result of insufficient sampling during the Fourier transforms, but upon increasing the kpoint sampling, the data do not appear to shed these features. On the other hand, taking the absolute value of these data appears to generate smoother results. Are the real and imaginary parts on their own well-represented and meaningful, or is it only the magnitude of the velocities that are?  If the latter, is there a way to recover the real and imaginary parts via some rotation or something similar? 

Of lesser importance is the finding that it appears that the real part of the matrix elements are anti-symmetric with respect to inversion, whereas the imaginary parts appear to be symmetric with inversion. Classically, I can imagine that anti-symmetric makes sense, but what makes the imaginary part symmetric instead? 

Assuming there aren't any catastrophic errors with the above, I also wanted to ask about obeying symmetries. In the above data for Aluminum, both the absolute values of the velocities and the eigenenergies along the kpoint paths agreed very very well. However things got more interesting when I tried a similar test for Iron, in particular when non-collinear spins were included in the calculation. In the same style of the above plot, here are the results for BCC Iron under different `spintype` conditions, using a 16x16x16 kpoint grid. 
![velocities](https://user-images.githubusercontent.com/32557266/131529023-699e4a31-c6bd-432a-ba3a-dcf244b150c1.png)

In the case of `spintype spin-orbit` and `spintype vector-spin`, the matrix elements are no longer perfectly (anti-) symmetric. Additionally, the eigenenergies are no longer perfectly symmetric. For example for the `vector-spin` calculation I found a maximum energy different between points along the two kpoint paths to be on the order of 40 meV. JDFTx indeed finds inversion symmetry in all three cases. I will also note that I'm using JDFTx v1.5 and 1.6, which I imagine might be relevant because there was a magnetic symmetry bug fix that happened with the release of v1.5. 

Are these issues related to simply the accuracy of the wannierization? If I use better windows and/or trial orbitals and/or kpoint sampling, do you expect that these will go away? It only is an issue for the cases of non-collinear spins (`spintype z-spin` gave no differences), so I wanted to check to see if there was something to worry about in the non-collinear situations in particular.  

Sorry for so many words. I hope my questions make sense and I would be very appreciative of any clarification. I'm happy to share any related files if they may be helpful. Thank you very much for your time! 

Regards,
Chris

**Comment by shankar1729** [comment]
Hey Chris,

Hope everything is well with you as well! The central issue you are encountering here is that the phase of each wavefunction is undefined. This does not affect diagonal matrix elements of any operator because the phase cancels between the bra and ket. However, at degenerate points / band crossings with a g-fold energy degeneracy, you have a g x g undetermined unitary transformation which does affect even the diagonal matrix elements (instead of the 1 x 1 unitary transformation = unknown phase).

The situation is completely different for off-diagonal matrix elements where adding a phase phi1 and phi2 to bands b1 and b2 changes the 1-2 matrix element by a phase phi1 - phi2. This has no effect on any physically meaningful quantity, but the complex matrix element itself is completely dependent on the phase of each column in the unitary matrix picked by the diagonalization algorithm. These algorithms tend to pick a "standard phase" that tries to keep the unitary matrix stable across different runs with slightly different numerics at round-off error for the same matrix, but do not guarantee any sort of continuity in phases for a family of continuously varying matrices (which is what we have for H(k)).

I suspect your observed symmetry / asymmetry for Al were also accidental. All I can say is that if the real part turned out symmetric due to the standard phase choice of the diagonalizer mentioned above, then the imaginary part should be antisymmetric. This is because if psi is an eigenstate at k, psi* is an eigenstate at -k. Consequently, k and -k are related by hermitian conjugates for matrix elements for an identical phase choice. In your case, you saw the real part antisymmetric in k, which means that the phase was pi between k and -k in addition to the conjugate (i.e. anti-hermitian). Therefore, your imaginary part turned out to be symmetric. Regardless, this is accidental and should not be relied on.

In short, this is a gauge freedom in the wavefunctions that does not have an easy fix if you want to examine phase of matrix elements. The best solution is to always plot a physically measurable quantity instead, such as the magnitude.

Best,
Shankar

**Comment by ciccarino** [comment]
Hey Shankar,

I see, I think this makes sense, thank you very much for your answer! I then have a follow up question. If I understand you correctly, the phase is definitely specific to the wavevector, band, and their associated wavefunction. However, can I get away with taking the product of two velocity matrix elements of different directions? In other words, does something like (v_x^{0,2})^* (v_y^{0,2}) also cancel the phases, like you mention in the diagonal matrix elements case? In this case I imagine that the phases are equivalent between v_x and v_y, and so upon multiplication (with one matrix element being complex conjugated), the phases can multiply to 1. I tried this with Al and I don't see any strange discontinuity behavior. Indeed when I do this, it is the imaginary part that is antisymmetric, and the real part is symmetric with respect to k. 

Also, I am still wondering about the second question I asked above, related to the symmetry of |v| with the inclusion of spin-orbit or vector-spin flags. Is my interpretation that the wannierization is simply not stringent enough a reasonable explanation for why |v| is not symmetric with respect to the symmetries of the unit cell?  Or is there another way to more rigorously capture the symmetries?

Thanks again for your time! 
Chris

**Comment by shankar1729** [comment]
Hi Chris,

Indeed, conjugated products of matrix elements (i.e. bilinears) are observables, just like the norm of the matrix elements and hence don't have the phase problem.

As for the symmetry question, I don't think it has to do with wannierization at all. In z-spin mode, spin polarization is completely decoupled with spatial degrees of freedom and you have a -x to +x symmetry in BCC iron. In spin-orbit mode, the spin is forced to be balanced, so there is no net magnetization and you have a -x to +x symmetry. However in vector-spin mode, you now have a net magnetization in some direction and that could be breaking the -x to +x symmetry in velocity by spin-orbit coupling. It may be worthwhile to look closer at the magnetization direction and the symmetries after the introduction of this magnetization.

Best,
Shankar
